libminifi/src/utils/NetworkInterfaceInfo.cpp

```diff
@@ -0,0 +1,158 @@
+  WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), -1, result_string.data(), size_needed, nullptr, nullptr);
+  return result_string;
+}
+}
+
+NetworkInterfaceInfo::NetworkInterfaceInfo(const IP_ADAPTER_ADDRESSES* adapter) {
+  name_ = utf8_encode(adapter->FriendlyName);
+  for (auto unicast_address = adapter->FirstUnicastAddress; unicast_address != nullptr; unicast_address = unicast_address->Next) {
+    if (unicast_address->Address.lpSockaddr->sa_family == AF_INET) {
+      ip_v4_addresses_.push_back(OsUtils::sockaddr_ntop(unicast_address->Address.lpSockaddr));
+    } else if (unicast_address->Address.lpSockaddr->sa_family == AF_INET6) {
+      ip_v6_addresses_.push_back(OsUtils::sockaddr_ntop(unicast_address->Address.lpSockaddr));
+    }
+  }
+  running_ = adapter->OperStatus == IfOperStatusUp;
+  loopback_ = adapter->IfType == IF_TYPE_SOFTWARE_LOOPBACK;
+}
+#else
+NetworkInterfaceInfo::NetworkInterfaceInfo(const struct ifaddrs* ifa) {
+  name_ = ifa->ifa_name;
+  if (ifa->ifa_addr->sa_family == AF_INET) {
+    ip_v4_addresses_.push_back(OsUtils::sockaddr_ntop(ifa->ifa_addr));
+  } else if (ifa->ifa_addr->sa_family == AF_INET6) {
+    ip_v6_addresses_.push_back(OsUtils::sockaddr_ntop(ifa->ifa_addr));
+  }
+  running_ = (ifa->ifa_flags & IFF_RUNNING);
+  loopback_ = (ifa->ifa_flags & IFF_LOOPBACK);
+}
+#endif
+
+namespace {
+struct HasName {
+  explicit HasName(const std::string& name) : name_(name) {}
+  bool operator()(const NetworkInterfaceInfo& network_interface_info) {
+    return network_interface_info.getName() == name_;
+  }
+  const std::string& name_;
+};
+}
+
+std::vector<NetworkInterfaceInfo> NetworkInterfaceInfo::getNetworkInterfaceInfos(std::function<bool(const NetworkInterfaceInfo&)> filter,
+                                                                                                     const utils::optional<uint32_t> max_interfaces) {
+  std::vector<NetworkInterfaceInfo> network_adapters;
+#ifdef WIN32
+  ULONG buffer_length = sizeof(IP_ADAPTER_ADDRESSES);
+  if (ERROR_BUFFER_OVERFLOW != GetAdaptersAddresses(0, 0, nullptr, nullptr, &buffer_length))
+    return network_adapters;
+  std::vector<char> bytes(buffer_length, 0);
+  IP_ADAPTER_ADDRESSES* adapter = reinterpret_cast<IP_ADAPTER_ADDRESSES*>(bytes.data());
+  if (NO_ERROR == GetAdaptersAddresses(0, 0, nullptr, adapter, &buffer_length)) {
+    while (adapter != nullptr) {
+      NetworkInterfaceInfo interface_info(adapter);
+      if (filter(interface_info)) {
+        auto it = std::find_if(network_adapters.begin(), network_adapters.end(), HasName(interface_info.getName()));
+        if (it == network_adapters.end()) {
+          network_adapters.emplace_back(std::move(interface_info));
+        } else {
+          interface_info.moveAddressesInto(*it);
+        }
+      }
+      if (max_interfaces.has_value() && network_adapters.size() >= max_interfaces.value())
+        return network_adapters;
+      adapter = adapter->Next;
+    }
+  }
```

Comment:
it could be useful to log an error if `GetAdaptersAddresses` returns something other than `NO_ERROR`

Reply:
Good idea, added log_error-s  in https://github.com/apache/nifi-minifi-cpp/pull/1116/commits/19f9c59dfa2700482fc26364c78cb6c1a0fc5822